File: internal/domain/placement.go
package domain

import "slices"

// This file is what a Run's stated objective does to a placement. The objective
// is public API, so "fastest_start" is a promise Mercator makes about where the
// Run lands, and until this existed it was a word that changed nothing: every
// candidate was ranked on one blended dollar score whose only time term was a
// weight nothing populated outside the balanced objective, so a Run that asked
// to start soonest was placed on whichever machine was a fraction of a cent
// cheaper and, when prices tied, on whichever offer ID sorted first.
//
// An objective is a ranking rather than an exchange rate. Converting a second
// of waiting into dollars needs a number nobody has measured, and inventing one
// per objective would be the unmeasured constant this codebase keeps deleting.
// What a Run did state is which quantity it wants least of, and that orders
// candidates on its own.

// BalancedWaitingUSDPerSecond is what the balanced objective presumes a second
// spent waiting to start is worth: 1.80 USD an hour, roughly the rent on the
// machine doing the waiting. It is a stated assumption rather than a
// measurement, said once here so the scheduler and the Lab's reference model
// cannot disagree about a number neither of them measured.
const BalancedWaitingUSDPerSecond = 0.0005

// Prefers reports whether this candidate is the better placement under the
// objective the Run stated. It is a total order: candidates that tie on every
// term the objective names fall back to the offer snapshot ID, so one offer set
// produces one decision however the offers arrived.
func (policy PlacementPolicy) Prefers(candidate, incumbent CandidateDecision) bool {
	if order := slices.Compare(policy.rank(candidate), policy.rank(incumbent)); order != 0 {
		return order < 0
	}
	return candidate.OfferSnapshotID < incumbent.OfferSnapshotID
}

// rank is what this objective orders candidates by, most significant term
// first. Every objective ends in the term another one leads with: a Run that
// asked for speed still takes the cheaper of two equally quick machines, and a
// Run that asked for price still takes the quicker of two equally cheap ones,
// which is the only thing that lets locality decide a placement between offers
// at one price.
//
// Completion is start plus the runtime the Run expects. That runtime is the
// same for every candidate today, so this ranks exactly as fastest_start does
// until something predicts per-candidate throughput, which is phase 4. It is
// written as the sum it means rather than as the shortcut it currently equals.
func (policy PlacementPolicy) rank(candidate CandidateDecision) []float64 {
	cost := candidate.ScoreUSD
	start := candidate.Estimates.StartSeconds.Expected
	switch policy.Objective {
	case ObjectiveFastestStart:
		return []float64{start, cost}
	case ObjectiveFastestCompletion:
		return []float64{start + policy.ExpectedRuntimeSeconds, cost}
	default:
		return []float64{cost, start}
	}
}

// SelectionReason names the rule that chose the winner, so the decision record
// says what it ranked on. A Run that asked for the earliest start and got the
// costliest machine is explained by its own objective, and recording that as
// LOWEST_SCORE would describe a comparison Mercator did not make.
func (policy PlacementPolicy) SelectionReason() string {
	switch policy.Objective {
	case ObjectiveFastestStart:
		return "EARLIEST_START"
	case ObjectiveFastestCompletion:
		return "EARLIEST_COMPLETION"
	default:
		return "LOWEST_SCORE"
	}
}

File: web/app/src/lib/placement.ts
import type { StoredCandidateDecision } from "@/lib/deployment/contracts";

/**
 * priced reports whether the dollars in a candidate's cost estimate are a price
 * somebody quoted. It reads the source of the cost estimate, which is where the
 * decision states that nobody did.
 */
export function priced(candidate: StoredCandidateDecision): boolean {
  return candidate.estimates?.cost_usd?.source !== "unpriced";
}

/**
 * orderCandidates is the reading order of a Booking Decision's candidates, and it
 * is the placement's own ranking rather than a sort on the score.
 *
 * The score is in dollars and a candidate nobody quoted has none of them, so its
 * score states only the waiting and the doubt. Sorting the rows on that number
 * put the unquoted machine at the top as the cheapest thing in the fleet, at a
 * fraction of the score of the machine Mercator actually selected, which is the
 * inverse of how the placement ranked them: an unpriced candidate is taken only
 * when nothing priced will do. The decision names that rule in its selection
 * reasons, and this is the same rule in the view of it.
 */
export function orderCandidates(
  candidates: StoredCandidateDecision[],
  selectedOfferId?: string,
): StoredCandidateDecision[] {
  return [...candidates].sort((a, b) => {
    if (a.offer_snapshot_id === selectedOfferId) return -1;
    if (b.offer_snapshot_id === selectedOfferId) return 1;
    if (a.feasible !== b.feasible) return a.feasible ? -1 : 1;
    if (priced(a) !== priced(b)) return priced(a) ? -1 : 1;
    const as = a.score_usd ?? Number.POSITIVE_INFINITY;
    const bs = b.score_usd ?? Number.POSITIVE_INFINITY;
    return as - bs;
  });
}
